pkg/sentry/fsimpl/fuse/read_write.go

```diff
@@ -197,11 +199,11 @@ func (fs *filesystem) Write(ctx context.Context, fd *regularFileFD, off uint64,
 			toWrite = maxWrite
 		}
 
-		in.Offset = off + uint64(written)
-		in.Size = toWrite
+		in.Header.Offset = off + uint64(written)
+		in.Header.Size = toWrite
+		in.Payload = data[written : written+toWrite]
```

Comment:
I think this is incorrect. In Linux, the payload is *not* part of the request headers and `fuse_in_header.len` doesn't include the payload len.

In linux, the write path begin in [`fuse_perform_write`](https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/file.c#L1107). You can see the actual write payload is passed along in `req->pages` in `fuse_fill_write_pages`. If you follow the call chain `fuse_perform_write` > `fuse_send_write_pages` > `fuse_send_write` > `fuse_send_request` > `__fuse_send_request` > [`queue_request`](https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/dev.c#L309), you can see `fuse_in_header.len` doesn't include the payload len.

Reply:
I'm not sure it's a correct assessment.
If you look into https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/dev.c#L309
```
	req->in.h.len = sizeof(struct fuse_in_header) +
		len_args(req->in.numargs, (struct fuse_arg *) req->in.args);
```
Here in `req->in.h` the struct is in fact `fuse_in_header`
See: https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/fuse_i.h#L300
```
struct fuse_req {
	/** This can be on either pending processing or io lists in
	    fuse_conn */
	struct list_head list;

	/** Entry on the interrupts list  */
	struct list_head intr_entry;

	/** refcount */
	atomic_t count;

	/** Unique ID for the interrupt request */
	u64 intr_unique;

	/* Request flags, updated with test/set/clear_bit() */
	unsigned long flags;

	/** The request input */
	struct fuse_in in;
```
And then https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/fuse_i.h#L165
```
/** The request input */
struct fuse_in {
	/** The request header */
	struct fuse_in_header h;
```
Overall I'm not sure how writes would work if `req` payload din't include the total length.
All functions that work on `req.in` are allocating full length of header + all the data. 
See https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/dev.c#L65 for example.

Reply:
> If you look into https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/dev.c#L309
> 
> 	req->in.h.len = sizeof(struct fuse_in_header) +
> 		len_args(req->in.numargs, (struct fuse_arg *) req->in.args);

On a closer look, I think you're right. I didn't notice that [here](https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/file.c#L955), `in.args[1].size` is the payload len, so `len_args(...)` includes it.

> Overall I'm not sure how writes would work if req payload din't include the total length.
> All functions that work on req.in are allocating full length of header + all the data.
> See https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/dev.c#L65 for example.

The payload len clearly has to be included somewhere. It is already a part of [`FUSE_WRITE_IN`](https://elixir.bootlin.com/linux/v4.10/source/fs/fuse/file.c#L946). Many servers don't look at `FUSE_IN_HEADER.len` and simply use the opcode to figure out which header to parse and whether that operation has a payload, for example see libfuse [here](https://github.com/libfuse/libfuse/blob/master/lib/fuse_lowlevel.c#L2547). However this is a moot point, since `FUSE_IN_HEADER.len` also includes the payload len.

--------------

Combining the payload into the write header as done in this PR seems more complicated than it needs to be. We could just do:

```
req := fs.conn.NewRequest(auth.CredentialsFromContext(ctx), uint32(t.ThreadID()), inode.nodeID, linux.FUSE_WRITE, &in)
req.payload = data[written : written+toWrite]
req.hdr.Len += len(req.payload)
``` 

Our FUSE device handles marshaling the payload contiguously after the headers, see [here](https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/fsimpl/fuse/dev.go;l=211;drc=9df539f9e893bf676b8f344a0c4d7909258940f0;bpv=1;bpt=1).

If you prefer to combine the payload with the write header, please remove [`Request.payload`](https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/fsimpl/fuse/request_response.go;l=100;drc=9df539f9e893bf676b8f344a0c4d7909258940f0) throughout, as it would become unused.

Reply:
Just to clarify we have no many available options.
Look into NewRequest, where FUSEHeaderIn.Len depends on `payload.SizeBytes()`.
```
func (conn *connection) NewRequest (..., payload marshal.Marshallable) *Request {
  hdrLen := (*linux.FUSEHeaderIn)(nil).SizeBytes()
  hdr := linux.FUSEHeaderIn{
		Len:    uint32(hdrLen + payload.SizeBytes()),
}
```
Wrapping data into marshallable payload means no changes in NewRequest function. But doing it in an old way require adding function argument like dataPayloadSize, so it would look like `Len:    uint32(hdrLen + payload.SizeBytes() + dataPayloadSize),`

> If you prefer to combine the payload with the write header, please remove [`Request.payload`](https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/fsimpl/fuse/request_response.go;l=100;drc=9df539f9e893bf676b8f344a0c4d7909258940f0) throughout, as it would become unused.

Ok, I will remove Requestr.payload and all the related code